### PR TITLE
Add listvms pagination test

### PR DESCRIPTION
**Pull request by misterdjules**
This is basically a reopen of #4, but for some reason GitHub doesn't allow me to reopen that one.

The changes are the same as #4, except for the last commit on this PR which includes an update to the node-moray dependency. That update includes a fix for node-fast, which fixes a problem where the node process would hang due to node-moray connections still open event after closing them.

### CODE REVIEW

**Comment by misterdjules** [comment]
Tested in nightly and tests pass, no process hung. /cc @trentm @richardkiene

File: test/vmapi.test.js [hunk 1]
```diff
@@ -8,24 +8,28 @@
...
 var libuuid = require('libuuid');
-function uuid() {
-    return (libuuid.create());
-}
-var util = require('util');
+var vasync = require('vasync');
+var moray = require('moray');
+var ldapjs = require('ldapjs');
+var assert = require('assert-plus');
+
```
Comment (trentm) [inline test/vmapi.test.js]: All these imports in alphabetical order please. `var util ...` as well.

File: test/vmapi.test.js [hunk 2]
```diff
@@ -61,6 +65,17 @@ var ROLE_TAG_TWO = '25a852d6-cf2c-11e3-a59f-77a5d70ae240';
...
+// These constants are used for the pagination tests.
+// Create a number of VMs that is large enough to make VMAPI use the pagination
+// logic when listing them. Currently, VMAPI sets the limit for the number of
+// entries that can be listed in one page to 1000.
+var LIST_VMS_PAGINATION_LIMIT = 1000;
+var NB_PAGINATION_TEST_VMS_TO_CREATE = LIST_VMS_PAGINATION_LIMIT * 2 + 1;
+
+var testPaginationVms;
+var leftoverTestPaginationVms;
```
Comment (trentm) [inline test/vmapi.test.js]: Two blank lines before a "section" (i.e. the "// --- ..." headings).

File: test/vmapi.test.js [hunk 3]
```diff
@@ -110,6 +125,116 @@ function waitForValue(fn, params, prop, value, callback) {
     return check();
 }
 
+function connectToMoray(callback) {
+    var MORAY_CLIENT_CONFIG = {
+        host: process.env.MORAY_IP || '10.99.99.17',
+        port: 2020,
+        log: new Logger({
+            name: 'moray',
```
Comment (trentm) [inline test/vmapi.test.js]: I wouldn't cry if you changed the bunyan usage to the more modern:

```
var bunyan = require('bunyan');
// ...
log: bunyan.createLogger({ ... });
```

File: package.json [hunk 4]
```diff
@@ -24,6 +24,8 @@
     "smartdc-auth": "2.1.6"
   },
   "devDependencies": {
+    "ldapjs": "^1.0.0",
```
Comment (trentm) [inline package.json]: Perhaps just dep on the smaller `ldap-filter` <https://github.com/pfmooney/node-ldap-filter>?

File: test/vmapi.test.js [hunk 5]
```diff
@@ -110,6 +125,116 @@ function waitForValue(fn, params, prop, value, callback) {
...
+                addVmsQueue.push(libuuid.create());
+            }
+
+            addVmsQueue.close();
+            addVmsQueue.on('end', function testVmsCreationDone() {
+                return next(null, morayClient);
+            });
+
+            function createTestVm(vmUuid, callback) {
+                var vmObject = {
```
Comment (trentm) [inline test/vmapi.test.js]: This is re-using the 'callback' var. 
I'm surprised this is passing 'make check'.

**Comment by trentm** [comment]
otherwise, LGTM.

**Comment by misterdjules** [comment]
Thanks, merged in 84d1ad332329476a7fa49a3ecf19d2567df3e4c7.

**Pull request closed without merging**